lib/flows/general/administrative.py

```diff
@@ -335,7 +335,7 @@ class Foreman(flow.WellKnownFlow):
   # How often we refresh the rule set from the data store.
   cache_refresh_time = 60
 
-  lock = threading.Lock()
+  lock = threading.RLock()
```

Comment:
This won't change anything, I suggest reverting this here. I have a fix for the race condition which I'll send soon.

What's happening here is that this lock is never used but overwritten in the constructor (of AFF4Object) with a PicklableLock. Turns out there is a  thread safety problem in the utils.PicklableLock (you had one job!! :) ) so under load multiple threads could create internal locks for the PickleableLock object at the same time (and enter the locked section at the same time).